### PR TITLE
Add train-test split for tabular data

### PR DESCRIPTION
**Pull request by AdrianM0**

### CODE REVIEW

File: data/tabular/train_test_split.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+REPRESENTATIONS = [
+    'SMILES',
+    'SELFIES',
+    'PSMILES',
+    'DeepSMILES',
+    'canonical SMILES',
+    'InChI',
+    'TUCAN',
+    'IUPAC name'
+    ]
```
Comment (kjappelbaum) [inline data/tabular/train_test_split.py:26]: why do you pass them all into the function (?) if the function only takes SMILES?

File: data/tabular/train_test_split.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    for path in csv_paths:
+        read_dataset = pd.read_csv(path)
+        try:
+            read_dataset = read_dataset.drop("split", axis=1)
+        except:
+            print("No split column")
+        
+        for col in read_dataset.columns:
+            if col in REPRESENTATIONS:
+                col_to_merge = col
```
Comment (kjappelbaum) [inline data/tabular/train_test_split.py:126]: Perhaps remove all the complexity with other representations and assume you have one column called "SMILES". That is what the sampling code assumes, too

File: data/tabular/train_test_split.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+
+        train_test_df = train_test_df.rename(columns={"REPR" : col_to_merge})
+        merged_data = pd.merge(read_dataset, train_test_df, on=col_to_merge, how="left")
+        merged_data = merged_data.dropna()
```
Comment (kjappelbaum) [inline data/tabular/train_test_split.py:130]: did you test this logic? Do you end up with only one split column in the end?

File: data/tabular/train_test_split.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+
+    Args:
+        s (str): the string to print
+    """
+    print(s, flush=True, file=sys.stderr)
+
+
+def create_scaffold_split(df, seed, frac, entity):
+    """create scaffold split. it first generates molecular scaffold for each molecule and then split based on scaffolds
+    reference: https://github.com/chemprop/chemprop/blob/master/chemprop/data/scaffold.py
```
Comment (kjappelbaum) [inline data/tabular/train_test_split.py:40]:
```suggestion
   adapted from https://github.com/chemprop/chemprop/blob/master/chemprop/data/scaffold.py
```

if one-to-one copy, say "copied from". Thanks for adding the reference!

File: data/tabular/train_test_split.py (hunk 1 above)
Comment (AdrianM0) [reply data/tabular/train_test_split.py]: was thinking of maybe conversion tools between the different representations?

File: data/tabular/train_test_split.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+
+def create_scaffold_split(df, seed, frac, entity):
+    """create scaffold split. it first generates molecular scaffold for each molecule and then split based on scaffolds
+    adapted from: https://github.com/mims-harvard/TDC/tdc/utils/split.py
+
+    Args:
+        df (pd.DataFrame): dataset dataframe
+        fold_seed (int): the random seed
+        frac (list): a list of train/valid/test fractions
+        entity (str): the column name for where molecule stores
```
Comment (MicPie) [inline data/tabular/train_test_split.py:33]: `entity` must be always a SMILES representation?

File: data/tabular/train_test_split.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            scaffold = MurckoScaffold.MurckoScaffoldSmiles(
+                mol=Chem.MolFromSmiles(smiles), includeChirality=False
+            )
+            scaffolds[scaffold].add(i)
+        except:
+            print_sys(smiles + " returns RDKit error and is thus omitted...")
+            error_smiles += 1
+
+    train, val, test = [], [], []
+    train_size = int((len(df) - error_smiles) * frac[0])
```
Comment (MicPie) [inline data/tabular/train_test_split.py:56]: Why is `error_smiles` subtracted from all three splits?

File: data/tabular/train_test_split.py (hunk 3 above)
Comment (kjappelbaum) [reply data/tabular/train_test_split.py:114]: can you perhaps add a small unit test for this script?

File: data/tabular/train_test_split.py (hunk 6 above)
Comment (AdrianM0) [reply data/tabular/train_test_split.py:56]: it is always subtracted from the length of the whole dataframe not from the individual splits length, to adjust the ratio of train-valid-test splits based on the ratio of error_smiles.

File: data/tabular/train_test_split.py (hunk 3 above)
Comment (AdrianM0) [reply data/tabular/train_test_split.py]: added a script

File: data/tabular/train_test_split.py (hunk 5 above)
Comment (AdrianM0) [reply data/tabular/train_test_split.py:34]: i fixed this now

File: data/tabular/test_split.py [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+    test_df["split"] = len(test_df) * ["test"]
+
+    # merge train and test across all datasets
+    merge = pd.concat([train_df, test_df], axis=0)
+    # rewrite data_clean.csv for each dataset
+    rewrite_data_with_splits(paths_to_data, merge)
+
+
+if __name__ == "__main__":
+    test_split(42)
```
Comment (kjappelbaum) [inline data/tabular/test_split.py:38]: thanks for adding this script but for a real test I would have expected some assert statements. But let me try it

File: data/tabular/test_split.py [hunk 8]
```diff
@@ -0,0 +1,38 @@
+# from ..train_test_split import create_scaffold_split, rewrite_data_with_splits
+import pandas as pd
+import importlib
+import os
+from glob import glob
+import sys
+from train_test_split import create_scaffold_split, rewrite_data_with_splits
+
+def test_split(seed):
+    paths_to_data = glob("tests/csvs_no_split_column/*.csv")
```
Comment (kjappelbaum) [inline data/tabular/test_split.py:10]: seems you did not commit this code

**Comment by kjappelbaum** [comment]
I have now run it and it seems okay to me. I added also some optional checks (enabled by default) that should help us spot if something goes wrong.

**Comment by kjappelbaum** [comment]
actually, there will probably be a bug due to the merge behavior if there are no SMILES in a given data frame

**Comment by kjappelbaum** [comment]
BTW, also made override a flag users must enable as I find it dangerous for this to be the default

File: data/tabular/train_test_split.py (hunk 6 above)
Comment (kjappelbaum) [reply data/tabular/train_test_split.py:72]: yep, this seems ok

Comment (MicPie) [reply data/tabular/train_test_split.py:72]: ah, yeah, make sense, sorry, I missed the entire df.

**Comment by MicPie** [comment]
> BTW, also made override a flag users must enable as I find it dangerous for this to be the default
The default could be to rename the original file and still keep it and then just name the new one `data_clean.csv`.

**Comment by kjappelbaum** [comment]
>The default could be to rename the original file and still keep it and then just name the new one data_clean.csv.

that is what the code now does.

File: data/tabular/train_test_split.py [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+
+            col_to_merge = "SMILES"
+            merged_data = pd.merge(
+                read_dataset, train_test_df, on=col_to_merge, how="left"
+            )
+            merged_data = merged_data.dropna()
+            if override:
+                merged_data.to_csv(path, index=False)
+            else:
+                merged_data.to_csv(path.replace(".csv", "_split.csv"), index=False)
```
Comment (MicPie) [inline data/tabular/train_test_split.py:161]: Copy `data_clean.csv` to something like `data_clean_default_split.csv` and save the new split info in `data_clean.csv`. Like that we don't have to rename the files or adapt the logic in the other scripts that are based on it.

**Pull request merged**